### PR TITLE
All color setters use lv_cpicker_set_hsv and return bool for change

### PR DESCRIPTION
**Pull request by paulpv**

### CODE REVIEW

**Comment by paulpv** [comment]
Not sure why this PR's commit history has the previous PR's commits in it.

Anyways, I'm looking in to figuring out why the actual set value doesn't match the displayed value (or vice versa).
It could be something as simple as the angle calculations, but those are the original code and at a casual glance those seem fine to me.

Again, I'm investigating, but please don't bother merging this, especially to 6.1, until this problem is fixed.

**Comment by kisvegabor** [comment]
@paulpv 
I've started to update the cpicker. Please give me a few days before sending further PRs to avoid conflicts.

**Comment by kisvegabor** [comment]
I've pushed a lot of changes to `feat-cpicker`. The main changes are:
- The preview area is removed. It can be added manually because it provides more flexibility.
- The indicator types are removed, only circle remained. I've checked some color pickers and found that it makes the most sens. 
- `style_indicator.body.padding` can make the indicator larger than the cpicker.
- `TYPE_RECT` uses `style_main.body.radius` to round the edges.
- `indic_colored` feature is added. It makes the indicator to show the current color. (it still can have a 
 border with any color)
- The `float` numbers are removed
- `TRI_OFFSET` is optimized
- Now the source code is ~1000 lines and IMO quite easy to follow

On STM32F429 Discovery (180MHz) it takes 41 ms (25 FPS) to draw a 200x200 color picker circle.
And it takes 4ms change only the selected point on the circle.

The test code is quite the same as what @paulpv  sent earlier:

```c
  lv_coord_t pickerSize = 200;

    static lv_style_t styleMain;
    lv_style_copy(&styleMain, &lv_style_plain);
    styleMain.line.width = 30;
    styleMain.body.radius = 5;

    static lv_style_t styleIndicator;
    lv_style_copy(&styleIndicator, &lv_style_pretty);
    styleIndicator.body.border.color = LV_COLOR_WHITE;
    styleIndicator.body.opa = LV_OPA_COVER;
    styleIndicator.body.border.opa = LV_OPA_COVER;

    lv_obj_t * scr = lv_scr_act();

    lv_obj_t * colorPicker = lv_cpicker_create(scr, NULL);
#if 1
    lv_obj_set_size(colorPicker, pickerSize, pickerSize);
    lv_cpicker_set_type(colorPicker, LV_CPICKER_TYPE_DISC);
#else
    lv_obj_set_size(colorPicker, pickerSize, pickerSize / 10);
    lv_cpicker_set_type(colorPicker, LV_CPICKER_TYPE_RECT);
#endif
    lv_obj_align(colorPicker, NULL, LV_ALIGN_CENTER, 0, 0);
    lv_cpicker_set_style(colorPicker, LV_CPICKER_STYLE_MAIN, &styleMain);
    lv_cpicker_set_style(colorPicker, LV_CPICKER_STYLE_INDICATOR, &styleIndicator);
    lv_cpicker_set_indic_colored(colorPicker, true);
```

Let me know what do you think.

**Comment by paulpv** [comment]
Was busy today off-site. I'll review tonight. Everything sounds good!

**Comment by paulpv** [comment]
I reverted my now mostly obsolete changed and merged yours.
That makes this PR have zero changes and now moot, but I will follow up here and close after any reply or a few days.

First impression that that I do like the code cleanup!
I am not familiar enough with your code base and would never have been that ballsy.

Second impression after seeing the new look is that there is a ton of dead white space.
I did like the original disc look more than the new look.
(I don't really care about the Rect and was just saving it and making it functional from @AloyseTech's original work for others that may like it better or want better performance).

My cheesy tests on an ESP32 SPI ILI9341 still show it only getting ~7-8 FPS.

I am 100% fine with your changes, but may want to absorb them a bit and offer future PRs.
With the basics in place, if there is anything I end up wanting to change for my personal project I can always work off of a fork and make it look more how I envision it...and who knows maybe PR some more changes.

You think your code is final and ready for merge w/ dev6.1?

**Comment by paulpv** [comment]
I am also not seeing that if I set the color at setup that is displays as set.
My code always shows the control starting colored as Green, never its default Red or any color that I set. :/

**Comment by kisvegabor** [comment]
> Second impression after seeing the new look is that there is a ton of dead white space.
I did like the original disc look more than the new look.
(I don't really care about the Rect and was just saving it and making it functional from @AloyseTech's original work for others that may like it better or want better performance).

We can add a `preview` option which draws the color inside the disk (as previously). However, it should be optional to allow users to put anything inside (e.g. a button being recolored). What do you think?

> I am 100% fine with your changes, but may want to absorb them a bit and offer future PRs.
With the basics in place, if there is anything I end up wanting to change for my personal project I can always work off of a fork and make it look more how I envision it...and who knows maybe PR some more changes.

Great! :slightly_smiling_face: 

> You think your code is final and ready for merge w/ dev6.1?

I think mainly yes. If there are no more questions/remarks it can be merged.

> I am also not seeing that if I set the color at setup that is displays as set.
My code always shows the control starting colored as Green, never its default Red or any color that I set. :/

It's working for me: 
`lv_cpicker_set_color(colorPicker, LV_COLOR_YELLOW);`

---

I've fixed the double click behavior as you [suggested](https://github.com/littlevgl/lvgl/commit/2600c1c3d97f0d879b7440d9f1251c6724d450b0#commitcomment-35321632).

**Comment by paulpv** [comment]
Nevermind; I had a stray `lv_cpicker_set_color` in my code that I thought I had deleted or at least commented; alas it was still there.

FYI: I am seeing some odd 565 conversion.
The following code in `lv_color.h` seems flawed and would never convert 565 red (0xF800) to 888 red (0xFF0000):
```
    ret.ch.red   = color.ch.red * 8;   /*(2^8 - 1)/(2^5 - 1) = 255/31 = 8*/
    ret.ch.green = color.ch.green * 4; /*(2^8 - 1)/(2^6 - 1) = 255/63 = 4*/
    ret.ch.blue  = color.ch.blue * 8;  /*(2^8 - 1)/(2^5 - 1) = 255/31 = 8*/
```
Searching the internets says that this is a more accurate conversion:
```
    ret.ch.red   = ( color.ch.red * 527 + 23 ) >> 6;
    ret.ch.green = ( color.ch.green * 259 + 33 ) >> 6;
    ret.ch.blue  = ( color.ch.blue * 527 + 23 ) >> 6;
```
I've figured out the magic numbers and can comment them in a PR that I may open on this.

**Comment by kisvegabor** [comment]
> I've figured out the magic numbers and can comment them in a PR that I may open on this.

Awesome! I know that it's perfect but didn't know about this fast running solution. A PR would be welcome!

**Comment by kisvegabor** [comment]
I've added `preview` feature.

**Comment by paulpv** [comment]
I'll review in a bit.
What seems odd about setting a color is that the default color is hsv 0,100,100, which looks Red.
But if I set LV_COLOR_RED, the method lv_color_rgb_to_hsv converts 565 red from 31,0,0 to hsv 0,255,31.
I don't know a ton about color conversion, but that A) looks like a bad S value (should be 0-100), B) I naively would think that the hsv conversion should convert a pure RGB color to have s and v of 100.

**Comment by paulpv** [comment]
I'll open an issue on concerns with `lv_color_rgb_to_hsv`, but basically:
1) should it be named `rgb8_to_hsv`? I am certain it can be made agnostic, but the `v = rgbMax` and `s = 255 *` make me think it is, as written, only valid for rgb888 values.
2) nothing constrains the s and v values to 100
3) nothing constrains uint16_t h from 0 to 359

**Comment by paulpv** [comment]
I'll also open an issue with concerns about `lv_color_to32`.
My tiny company moved yesterday and today and we still don't have internet at work, so my responses may be delayed and terse.
Sorry if I seem to be nitpicking your most excellent library; these seem like fundamental areas that don't seem to me to be working as I would expect them too.

**Comment by paulpv** [comment]
My laptop is still offline so apologies for continuing to discuss this here more convenient to me from my phone than not yet opening a new issue.

Please keep in mind that I am very new to this excellent library and never it existed before two weeks ago.

How many people do you suspect use this library's hsv code?
Being new to this library, and thinking the current code seems a bit broken, I propose that, to make this work better with microcontrollers, the hsv code be re-written to be similar to FastLED and use 0 to 255 for all H, S, and V values.
The best RGB to HSV conversions I have seen use a floating point 0.0 to 1.0 for all 3 values.
0 to 255 would be a similar normalized range.
This would also make the code color depth agonistic.
The current method signature could be kept for compatibility (even though it seems flawed to me), but a new method that handles normalized values would be added, as well as helper functions that will normalize the values, as well as an agnostic method that will automatically normalize the values.

Of course, if we have lots of methods to convert RGB to HSV, we would need the same number to convert HSV back to different RGB formats.

**Comment by embeddedt** [comment]
If the code really is broken (very likely, since most people probably just use RGB) and we kept the color scales the same, I think fixing the HSV code could be done in a minor release, provided that it keeps roughly the same color shade. I think users will tolerate the difference in color shades in exchange for proper HSV.

If we change the ranges then I think it has to be postponed to a major release, or we have to add two APIs (which I would really like to avoid for pollution reasons).

**Comment by paulpv** [comment]
I am finally back online at work!
Reviving this PR by adding some actual code changes! :)

I have comments in my local code about the separate `lv_color_rgb_to_hsv` and `lv_color_to32` issues.
I won't forget about the issues and I will either open new issues or new PRs once I determine both if there is indeed an issue and a good fix that I find.

**Comment by paulpv** [comment]
Heh! `+192 −191` Saved a line of code! :)

**Comment by embeddedt** [comment]
Looks good! I don't particularly like how the handle is now slightly larger than the color ring. It looks odd especially against a white background. I preferred the older style where the handle was the same size as the ring. But other than that, it's much faster and looks the same (minus the preview area)!

I found an issue with making the center background transparent (no preview):

```c
    styleMain.body.main_color = styleMain.body.grad_color = LV_COLOR_TRANSP;
```

The center is green instead.

File: src/lv_objx/lv_cpicker.c [hunk 1]
```diff
@@ -200,88 +200,79 @@ void lv_cpicker_set_style(lv_obj_t * cpicker, lv_cpicker_style_t type, lv_style_
...
 /**
  * Set the current value of a colorpicker.
  * @param cpicker pointer to colorpicker object
  * @param val current selected value [0..100]
+ * @return true if changed, otherwise false
  */
-void lv_cpicker_set_value(lv_obj_t * cpicker, uint8_t val)
+bool lv_cpicker_set_value(lv_obj_t * cpicker, uint8_t val)
 {
-    LV_ASSERT_OBJ(cpicker, LV_OBJX_NAME);
```
Comment (kisvegabor) [inline src/lv_objx/lv_cpicker.c:246]: These asserts should remain.

**Comment by kisvegabor** [comment]
@paulpv 
Good to hear that you successfully moved! :slightly_smiling_face: 

Fixing the hsv functions would be very welcome. Let's keep the current ranges for compatibility. Can you open a new PR for it?

@embeddedt 
>  don't particularly like how the handle is now slightly larger than the color ring.

It comes from the indicator's style's padding. Setting it the paddings to 0 will produce a "ring sized" knob.

> styleMain.body.main_color = styleMain.body.grad_color = LV_COLOR_TRANSP;

`LV_COLOR_TRANS` is used only on Chroma keyed images. To add transparency to the middle area, you should use `body.opa`. However, I don't recommend it because the middle area masks out inner parts of the triangles (the ring is built from triangles drawn between the center and the edge)

File: src/lv_objx/lv_cpicker.c (hunk 1 above)
Comment (paulpv) [reply src/lv_objx/lv_cpicker.c:246]: It is retained in the first method to access the object

**Comment by paulpv** [comment]
> > styleMain.body.main_color = styleMain.body.grad_color = LV_COLOR_TRANSP;
>
> LV_COLOR_TRANS is used only on Chroma keyed images. To add transparency to the middle area, you should use body.opa. However, I don't recommend it because the middle area masks out inner parts of the triangles (the ring is built from triangles drawn between the center and the edge)

I am not seeing anywhere that LV_COLOR_TRANS* is used in this PR.

**Comment by paulpv** [comment]
> > styleMain.body.main_color = styleMain.body.grad_color = LV_COLOR_TRANSP;
>
> LV_COLOR_TRANS is used only on Chroma keyed images. To add transparency to the middle area, you should use body.opa. However, I don't recommend it because the middle area masks out inner parts of the triangles (the ring is built from triangles drawn between the center and the edge)

I fake the transparency by setting the body main_color & grad_color to the theme background:
```
    lv_theme_t * th = lv_theme_get_current();
    styleMain.body.main_color = th->style.bg->body.main_color;
    styleMain.body.grad_color = styleMain.body.main_color;
```

**Comment by paulpv** [comment]
This PR is ready for final review.
I think it is a better looking compromise between @AloyseTech 's original concept and @kisvegabor 's latest.

I am satisfied w/ the color behavior with the understanding that something unrelated in the HSV code is causing the undesired results of setting 565 colors looking bad.
I will PR fixes in that area separately.

Obviously, since this control relies fundamentally on HSV, until those HSV issues are resolved this control may appear to have issues.

File: src/lv_objx/lv_cpicker.c (hunk 1 above)
Comment (paulpv) [reply src/lv_objx/lv_cpicker.c:246]: If you want me to add them back I will, but it seemed redundant when not accessed locally but further down the call stack.

Comment (embeddedt) [reply src/lv_objx/lv_cpicker.c:246]: I'm in favor of leaving the assert out here. The function only uses `lv_cpicker_get_hsv` and `lv_cpicker_set_hsv` so the object pointer is in fact already checked twice.

**Comment by paulpv** [comment]
Odd! (to me)
I have been instrumenting the code with logs to tweak a few nickpicks (later PRs).
Incrementing a single value (ex: Hue) by one causes 15 calls to `draw_disc_grad`:
```
15:28:04.382 -> refr_indic_pos hsv={269, 100, 100}
15:28:04.382 -> refr_indic_pos angle=269
15:28:04.382 -> invalidate_indic
15:28:04.382 -> 
15:28:04.382 -> draw_disc_grad start_angle=33, end_angle=60
15:28:04.382 -> draw_disc_grad start_angle=126, end_angle=231
15:28:04.382 -> draw_disc_grad start_angle=123, end_angle=237
15:28:04.382 -> draw_disc_grad start_angle=114, end_angle=243
15:28:04.486 -> draw_disc_grad start_angle=108, end_angle=249
15:28:04.486 -> draw_disc_grad start_angle=102, end_angle=258
15:28:04.486 -> draw_disc_grad start_angle=93, end_angle=267
15:28:04.486 -> draw_disc_grad start_angle=0, end_angle=357
15:28:04.486 -> draw_disc_grad start_angle=267, end_angle=450
15:28:04.486 -> draw_disc_grad start_angle=276, end_angle=441
15:28:04.486 -> draw_disc_grad start_angle=285, end_angle=432
15:28:04.486 -> draw_disc_grad start_angle=291, end_angle=426
15:28:04.486 -> draw_disc_grad start_angle=297, end_angle=420
15:28:04.523 -> draw_disc_grad start_angle=303, end_angle=414
15:28:04.523 -> draw_disc_grad start_angle=309, end_angle=408
```

I think this can be optimized.

File: src/lv_objx/lv_cpicker.c (hunk 1 above)
Comment (kisvegabor) [reply src/lv_objx/lv_cpicker.c:246]: @embeddedt  
Alright, leave it out. As there is nothing else in these functions just other functions calls (which have asserts it really should be fine).

File: src/lv_objx/lv_cpicker.c [hunk 2]
```diff
@@ -772,8 +740,39 @@ static void draw_indic(lv_obj_t * cpicker, const lv_area_t * mask, lv_opa_t opa_
...
+    lv_area_t indic_area = get_indic_area(cpicker, style_main, style_indic, ext);
+
+    lv_inv_area(lv_obj_get_disp(cpicker), &indic_area);
+}
+
+static lv_area_t get_indic_area(lv_obj_t * cpicker,
+                                const lv_style_t * style_main,
+                                const lv_style_t * style_indic,
+                                const lv_cpicker_ext_t * ext)
+{
```
Comment (kisvegabor) [inline src/lv_objx/lv_cpicker.c:748]: I'd prefer to pass only `cpicker`. It should be very fast to get the `styles` and `ext`.

**Comment by kisvegabor** [comment]
> Incrementing a single value (ex: Hue) by one causes 15 calls to draw_disc_grad:

What is the size of your display buffer? If it's small LittlevGL draws the object in multiple pieces. See [Drawing](https://docs.littlevgl.com/en/html/overview/drawing.html)

**Comment by paulpv** [comment]
```
// in lv_conf.h
#define LV_HOR_RES_MAX          (320)
#define LV_VER_RES_MAX          (240)

// in app code
static lv_color_t buf_1[LV_HOR_RES_MAX * 10];
static lv_color_t buf_2[LV_HOR_RES_MAX * 10];
lv_disp_buf_init(&disp_buf, buf_1, buf_2, LV_HOR_RES_MAX * 10);
```

**Comment by kisvegabor** [comment]
If the cpicker has 200x200 size then it has 40,000 px. Your buffer has 10x320 = 3,200 px. So the cpicker can be drawn in 40,000 / 3,200 = 12,5 (13) pieces.  

You can see it if you print the coordinates in `disp_flush`.

File: src/lv_objx/lv_cpicker.c (hunk 2 above)
Comment (paulpv) [reply src/lv_objx/lv_cpicker.c:748]: Fixed

**Review by kisvegabor — APPROVED** [review]
Looks good to me! Thank you!

**Comment by embeddedt** [comment]
Merged.

**Comment by embeddedt** [comment]
Something strange has happened. With the latest `feat-cpicker` branch (on our repo) the knob is not getting cleared when I move it (leaving a bunch of old knobs on the color picker ring).

Perhaps something got commented out during debugging?

**Comment by paulpv** [comment]
I did change the code to only invalidate if the value changes. Perhaps I introduced a bug, but "it works fine on my machine". I'll look at very soon (but not immediately).

**Comment by embeddedt** [comment]
>I did change the code to only invalidate if the value changes.

The value is definitely changing (because I can move the knob and the color inside changes as I move it). It seems like the background doesn't get redrawn.

**Comment by embeddedt** [comment]
I did a bit of debugging just now.

I made `invalidate_indic` invalidate the whole color picker (with `lv_obj_invalidate`). That worked, and also caused significant loss in performance.

Now I'm trying to find out why the background doesn't redraw even though that area is invalidated.

**Comment by paulpv** [comment]
I am not seeing this issue on my branch, which appears to be a bit behind your repo.
Here is how I create the control:
```
    const uint32_t dispWidth = lv_disp_get_hor_res(disp);
    const uint32_t dispHeight = lv_disp_get_ver_res(disp);
    pickerSize = (uint32_t)round(LV_MATH_MIN(dispWidth, dispHeight) * 0.8);

    static lv_style_t styleMain;
    lv_style_copy(&styleMain, &lv_style_plain);
    static lv_style_t styleIndicator;
    lv_style_copy(&styleIndicator, &lv_style_pretty);

    lv_theme_t * th = lv_theme_get_current();
    styleMain.body.main_color = th->style.bg->body.main_color;
    styleMain.body.grad_color = styleMain.body.main_color;
    styleMain.line.width = pickerSize * 0.2;
    //styleMain.body.radius = 5;

    styleIndicator.body.border.color = LV_COLOR_WHITE;
    styleIndicator.body.opa = LV_OPA_COVER;
    styleIndicator.body.border.opa = LV_OPA_COVER;

    lv_obj_t * scr = lv_disp_get_scr_act(disp);

    colorPicker = lv_cpicker_create(scr, NULL);
    lv_obj_set_size(colorPicker, pickerSize, pickerSize);
    lv_cpicker_set_type(colorPicker, LV_CPICKER_TYPE_DISC);
    lv_obj_align(colorPicker, NULL, LV_ALIGN_CENTER, 0, 0);
    lv_cpicker_set_style(colorPicker, LV_CPICKER_STYLE_MAIN, &styleMain);
    lv_cpicker_set_style(colorPicker, LV_CPICKER_STYLE_INDICATOR, &styleIndicator);
    lv_obj_set_event_cb(colorPicker, colorPicker_event_handler);
    lv_cpicker_set_indic_colored(colorPicker, true);      
    lv_cpicker_set_preview(colorPicker, true);      
```

**Comment by paulpv** [comment]
FYI: I am definitely seeing different behavior when I merge w/ the latest commit versus when I revert back to my last known working commit.
That said, I am having a problem w/ my [what I thought quality] K&H AD-12 breadboard introducing erroneous GPIO readings, so currently all bets are off when it comes to runtime behavior on my side. :/

**Comment by embeddedt** [comment]
@paulpv What happens if you start with your working commit, and merge `dev-6.1` in (but not `feat-cpicker`)? Does it still break?

If it does, then we must have broken something inside LittlevGL itself.

**Comment by paulpv** [comment]
Picking this back up now.

**Comment by paulpv** [comment]
The weirdness I was seeing was an unrelated change of GPIO pins.
Seems like the invalidate problem is due to something introduce in my  34faac98a734c73a3be6ee46bd897f267924a349 commit.

**Comment by kisvegabor** [comment]
I've just merged `feat-cpicker` to `dev-6.1`

**Pull request merged**